### PR TITLE
ddl: forbid type conversion between varchar and non-varchar

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/23624

Problem Summary:

The index value format is different between varchar and non-varchar strings. Thus, we need to reorganize index data if we support the change between `varchar` and non-`varchar`. Before column type change feature GA, let's forbid such changes.

### What is changed and how it works?

What's Changed: Omitted.

How it Works: Omitted.

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Breaking backward compatibility

### Release note 

- Forbid type conversion between varchar and non-varchar.

### CODE REVIEW

**Comment by AilinKid** [comment]
For now, it's a little bit ugly here, we should refine it when we do column type change. @tangenta

File: types/etc.go [hunk 1]
```diff
@@ -122,6 +122,12 @@ func IsString(tp byte) bool {
 	return IsTypeChar(tp) || IsTypeBlob(tp) || IsTypeVarchar(tp) || IsTypeUnspecified(tp)
 }
 
+// IsNonVarcharString returns a boolean indicating
+// whether the field type is a non-varchar string type.
+func IsNonVarcharString(tp byte) bool {
```
Comment (AilinKid) [inline types/etc.go:127]: any usage?

**Comment by wjhuang2016** [comment]
image
Even if not index is used, varchar -> var may be wrong.

File: executor/distsql.go [hunk 2]
```diff
@@ -1050,6 +1050,8 @@ func (w *tableWorker) compareData(ctx context.Context, task *lookupTableTask, ta
 				tp := &col.FieldType
 				idxVal := idxRow.GetDatum(i, tp)
 				tablecodec.TruncateIndexValue(&idxVal, w.idxLookup.index.Columns[i], col.ColumnInfo)
+				idxVal.SetCollation(charset.CollationBin)
```
Comment (zimulala) [inline executor/distsql.go:1053]: Please add a comment for here?

Comment (tangenta) [reply executor/distsql.go:1053]: No, we will remove this because there are some tests fails and it needs further investigation.

Comment (zimulala) [reply executor/distsql.go]: OK.

**Comment by tangenta** [comment]
/run-common-test=pr/1177

**Pull request merged**